src/main/java/org/jenkinsci/plugins/workflow/graphanalysis/AbstractFlowScanner.java

```diff
@@ -0,0 +1,344 @@
+     * @return True if we can have nodes to work with, otherwise false
+     */
+    public boolean setup(@CheckForNull Collection<FlowNode> heads, @CheckForNull Collection<FlowNode> blackList) {
+        if (heads == null || heads.size() == 0) {
+            return false;
+        }
+        Collection<FlowNode> fastEndNodes = convertToFastCheckable(blackList);
+        LinkedHashSet<FlowNode> filteredHeads = new LinkedHashSet<FlowNode>(heads);
+        filteredHeads.removeAll(fastEndNodes);
+
+        if (filteredHeads.size() == 0) {
+            return false;
+        }
+
+        reset();
+        myBlackList = fastEndNodes;
+        setHeads(filteredHeads);
+        return true;
+    }
+
+    /**
+     * Helper: version of {@link #setup(Collection, Collection)} where we don't have any nodes to blacklist
+     */
+    public boolean setup(@CheckForNull Collection<FlowNode> heads) {
+        if (heads == null) {
+            return false;
+        }
+        return setup(heads, Collections.EMPTY_SET);
+    }
+
+    /**
+     *  Helper: version of {@link #setup(Collection, Collection)} where we don't have any nodes to blacklist, and have just a single head
+     */
+    public boolean setup(@CheckForNull FlowNode head, @CheckForNull Collection<FlowNode> blackList) {
+        if (head == null) {
+            return false;
+        }
+        return setup(Collections.singleton(head), blackList);
+    }
+
+    /**
+     * Helper: version of {@link #setup(Collection, Collection)} where we don't have any nodes to blacklist and have just a single head
+     */
+    public boolean setup(@CheckForNull FlowNode head) {
+        if (head == null) {
+            return false;
+        }
+        return setup(Collections.singleton(head), Collections.EMPTY_SET);
+    }
+
+    /** Reset internal state so that we can begin walking a new flow graph
+     *  Public APIs need to invoke this before searches */
+    protected abstract void reset();
+
+    /**
+     * Set up to begin flow scanning using the filteredHeads as starting points
+     *
+     * This method makes several assumptions:
+     *
+     *   - {@link #reset()} has already been invoked to reset state
+     *   - filteredHeads has already had any points in {@link #myBlackList} removed
+     *   - none of the filteredHeads are null
+     * @param filteredHeads Head nodes that have been filtered against blackList
+     */
+    protected abstract void setHeads(@Nonnull Collection<FlowNode> filteredHeads);
+
+    /**
+     * Actual meat of the iteration, get the next node to visit, using & updating state as needed
+     * @param current Current node to use in generating next value
+     * @param blackList Nodes that are not eligible for visiting
+     * @return Next node to visit, or null if we've exhausted the node list
+     */
+    @CheckForNull
+    protected abstract FlowNode next(@Nonnull FlowNode current, @Nonnull Collection<FlowNode> blackList);
```

Comment:
Wh would we need it here since we already pass the blacklist in the setup() method?